### PR TITLE
Feature to retry failed download/install operations

### PR DESCRIPTION
**Pull request by ranasaria**
This change adds a configurable feature to retry download and install operations. The configuration of whether to do retry and how many times with what delays are specified in the config file with a new optional field called 'retry'. 

Also took this opportunity to add code to cleanup all downloaded after execution of script is done in all error and success paths.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "http-proxy-agent": "^2.1.0",
     "https-proxy-agent": "^2.2.1",
     "mkdirp": "^0.5.1",
+    "promise-retry": "^1.1.1",
```
Comment (anthonydresser) [inline package.json]: Why do you need a library to do this?

File: src/httpClient.ts [hunk 2]
```diff
@@ -19,10 +19,10 @@ export class HttpClient {
...
     public readonly eventEmitter = new EventEmitter({ wildcard: true });
 
-   /*
-    * Downloads a file and stores the result in the temp file inside the package object
-    */
-    public downloadFile(urlString: string, pkg: IPackage, proxy?: string, strictSSL?: boolean): Promise<void> {
+    /*
+     * Downloads a file and stores the result in the temp file inside the package object
+     */
+    public async downloadFile(urlString: string, pkg: IPackage, proxy?: string, strictSSL?: boolean): Promise<void> {
```
Comment (anthonydresser) [inline src/httpClient.ts]: why are you making this async if there is no await in the body?

File: src/interfaces.ts [hunk 3]
```diff
@@ -32,6 +32,15 @@ export interface IConfig {
...
     strictSSL: boolean;
     executableFiles: Array<string>;
+    /**
+     * Optional configuration for retries.
+     * Enabled flag is used to turn it on or off. It is off by default.
+     * Options object is a pass through configuration to the http://npmjs.org/retry module.
+     */
+    retry?: {
+        enabled: boolean,
+        options: any
```
Comment (anthonydresser) [inline src/interfaces.ts]: These should be typed, don't use any. I would also say you don't need the enabled property, if they pass a config option for retry its fair to assume they want to be doing retry.

File: package.json (hunk 1 above)
Comment (ranasaria) [reply package.json]: Why should we reimplement the code that a library has it and has been used by a lot of other libraries out there so has withstood the test of time.

File: src/httpClient.ts (hunk 2 above)
Comment (ranasaria) [reply src/httpClient.ts]: it does not hurt to have methods that return a Promise be marked async. I find that it improves readability. But I do not have a strong opinion about that.

File: src/interfaces.ts (hunk 3 above)
Comment (ranasaria) [reply src/interfaces.ts]: They were any because this is a pass through parameter to external library. If we make it typed in future when the details of the type changes we will need to make a correspondig change. However, in retrospect most folks would only add options so I am fine to make it strongly typed. The big advantage would be that I would be able to add details comments about each option with the retry configuration.

File: src/httpClient.ts (hunk 2 above)
Comment (ranasaria) [reply src/httpClient.ts]: See here for exampls of async method that do not have awaits in their bodies.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

File: src/httpClient.ts [hunk 4]
```diff
@@ -73,20 +73,20 @@ export class HttpClient {
...
             options = httpsOptions;
         }
 
         return options;
     }
 
-   /*
-    * Calculate the download percentage and stores in the progress object
-    */
+    /*
```
Comment (ranasaria) [inline src/httpClient.ts]: These were just changes done by 'Format Document'

File: src/serviceDownloadProvider.ts [hunk 5]
```diff
@@ -77,38 +77,46 @@ export class ServiceDownloadProvider {
...
+        const pkg: IPackage = {
+            installPath: installDirectory,
+            url: urlString,
+            tmpFile: undefined
+        };
+
+        const downloadAndInstall: () => Promise<void> = async () => {
+            try {
+                pkg.tmpFile = await this.createTempFile(pkg);
+				console.info(`\tdownloading the package: ${pkg.url}`);
```
Comment (ranasaria) [inline src/serviceDownloadProvider.ts]: fix spacing.

File: src/serviceDownloadProvider.ts [hunk 6]
```diff
@@ -77,38 +77,46 @@ export class ServiceDownloadProvider {
...
-                        resolve(true);
-                    }).catch(installError => {
-                        reject(installError);
-                    });
-                }).catch(downloadError => {
-                    reject(downloadError);
-                });
-            });
-        });
+        if (this._config.retry) {
```
Comment (ranasaria) [inline src/serviceDownloadProvider.ts]: do not need this conditional as when this._config.retry is undefined, withRetry can be set to do number of 'retries' as 0.

File: src/httpClient.ts (hunk 2 above)
Comment (anthonydresser) [reply src/httpClient.ts]: I know how async methods work. When using typescript they get interpreted different for example this code 
````typescript
    public getOrDownloadServer(): Promise<string> {
        return this.getServerPath().then(result => {
            if (result === undefined) {
                return this.downloadServerFiles().then(downloadResult => {
                    return downloadResult;
                });
            } else {
                return result;
            }
        });
    }
````
becomes
```` javascript
getOrDownloadServer() {
        return this.getServerPath().then(result => {
            if (result === undefined) {
                return this.downloadServerFiles().then(downloadResult => {
                    return downloadResult;
                });
            }
            else {
                return result;
            }
        });
    }
````
however this code
````typescript
public async getOrDownloadServer(): Promise<string> {
        return this.getServerPath().then(result => {
            if (result === undefined) {
                return this.downloadServerFiles().then(downloadResult => {
                    return downloadResult;
                });
            } else {
                return result;
            }
        });
    }
````
becomes
````javascript
getOrDownloadServer() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getServerPath().then(result => {
                if (result === undefined) {
                    return this.downloadServerFiles().then(downloadResult => {
                        return downloadResult;
                    });
                }
                else {
                    return result;
                }
            });
        });
    }
````

so just by adding an async identifer, the javascript becomes more complicated even though they are functionally identical. The return of a Promise already tells you its async, adding async just makes the resulting javascript more complicated unnecessarily.

File: src/interfaces.ts (hunk 3 above)
Comment (anthonydresser) [reply src/interfaces.ts]: Having any means that users have to explicitly know what retry library we are using and then go look up theirs options. This kind of thing is what typescript is meant to solve. we can include types for the underlying library.  https://www.npmjs.com/package/@types/promise-retry and just pull the option types from them. But there is also a chance we want to only allow a subset of the options for our application.

File: package.json (hunk 1 above)
Comment (anthonydresser) [reply package.json]: node module bloat is a real issue. Every node module has to be provably worth it. I took a look at this library (and really this is just a wrapper for a different library) and the real advantage is the exponentiation pull back feature https://github.com/tim-kos/node-retry . Do we need this? other wise a simple while loop with timeouts works just as well.

Comment (ranasaria) [reply package.json]: so are you with me using: https://github.com/tim-kos/node-retry with promisify instead?

File: src/httpClient.ts (hunk 2 above)
Comment (ranasaria) [reply src/httpClient.ts]: ok will remove async.

File: package.json (hunk 1 above)
Comment (ranasaria) [reply package.json]: promise-retry is just a wapper on https://github.com/tim-kos/node-retry, i can code directly using https://github.com/tim-kos/node-retry if you like but I would be duplicating a bunch of code in promise-retry.

Comment (ranasaria) [reply package.json]: I will leave this as is becasue service-downloader is a small project and we do not have module bloat issues here.

Comment (anthonydresser) [reply package.json]: I mean using the promise retry wrapper is fine, if we care about the features of the underlying library. Im not making a decision about whether we want the exponential retry, but I want that to be considered, because, like a said, a simple retry can be performed by just a while loops and settimeouts....

Comment (anthonydresser) [reply package.json]: module bloat has nothing to do with how small of a project it is....its about having a guideline you use when considering using libraries and then following that guideline....

Comment (ranasaria) [reply package.json]: yeah i know, in my experience when trying to address issues of failures around network flakiness using exponential backoff based retry strategy works best.

Comment (anthonydresser) [reply package.json]: Then using this library is fine, we can keep it, but add the typing files so the compiler can make sure we are using it correctly.

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "service-downloader",
-  "version": "0.1.6",
+  "version": "0.1.7",
```
Comment (anthonydresser) [inline package.json]: generally you only bump the version during doing the release, not when adding features.

File: src/serviceDownloadProvider.ts [hunk 8]
```diff
@@ -20,110 +20,136 @@ import { PlatformNotSupportedError, DistributionNotSupportedError } from './erro
 */
 export class ServiceDownloadProvider {
 
-    private httpClient = new HttpClient();
-    public readonly eventEmitter = new EventEmitter({ wildcard: true });
+	private httpClient = new HttpClient();
```
Comment (Charles-Gagnon) [inline src/serviceDownloadProvider.ts]: >	 [](start = 0, length = 1)

Why did you change these to tabs? Now there's mixed indentation spacing in this file

File: src/serviceDownloadProvider.ts [hunk 9]
```diff
@@ -20,110 +20,136 @@ import { PlatformNotSupportedError, DistributionNotSupportedError } from './erro
...
+
+	private install(pkg: IPackage): Promise<void> {
+		this.eventEmitter.emit(Events.INSTALL_START, pkg.installPath);
+		return decompress(pkg.tmpFile.name, pkg.installPath).then(() => {
+			this.eventEmitter.emit(Events.INSTALL_END);
+		});
+	}
+
+	private async withRetry(promiseToExecute: () => Promise<any>, retryOptions: IRetryOptions = { retries: 0 }): Promise<any> {
+		const promiseRetry = require('promise-retry');
```
Comment (Charles-Gagnon) [inline src/serviceDownloadProvider.ts]: This doesn't need to be a class member

File: src/serviceDownloadProvider.ts [hunk 10]
```diff
@@ -20,110 +20,136 @@ import { PlatformNotSupportedError, DistributionNotSupportedError } from './erro
...
+		// wrap function execution with a retry promise
```
Comment (Charles-Gagnon) [inline src/serviceDownloadProvider.ts]: Why are you doing this inline vs just importing it at the top?

File: package.json (hunk 1 above)
Comment (ranasaria) [reply package.json]: I am going to replace promise-retry with async-retry as I like its implementation much better (code you write to it is also more readable) and will add typings for that.

File: src/serviceDownloadProvider.ts (hunk 9 above)
Comment (anthonydresser) [reply src/serviceDownloadProvider.ts]: ^ also use import over require

File: src/serviceDownloadProvider.ts (hunk 10 above)
Comment (ranasaria) [reply src/serviceDownloadProvider.ts]: changed to import at the top.

File: src/serviceDownloadProvider.ts [hunk 11]
```diff
@@ -77,35 +77,41 @@ export class ServiceDownloadProvider {
...
+                pkg.tmpFile = await this.createTempFile(pkg);
+                console.info(`\tdownloading the package: ${pkg.url}`);
+                console.info(`\t                to file: ${pkg.tmpFile.name}`);
+                await this.httpClient.downloadFile(pkg.url, pkg, proxy, strictSSL);
+                console.info(`\tinstalling the package from file: ${pkg.tmpFile.name}`);
+                await this.install(pkg);
+
+            } finally {
+                // remove the downloaded package file
+                if (fs.existsSync(pkg.tmpFile.name)) {
```
Comment (anthonydresser) [inline src/serviceDownloadProvider.ts]: Don't use sync functions. use `promifisy` from `utils` to make exists and unlink into a Promise. https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original

File: src/serviceDownloadProvider.ts [hunk 12]
```diff
@@ -11,10 +11,10 @@ import { EventEmitter2 as EventEmitter } from 'eventemitter2';
 import * as tmp from 'tmp';
 
 import { Runtime, getRuntimeDisplayName } from './platform';
-import { IConfig, IPackage, Events } from './interfaces';
+import { IConfig, IPackage, Events, IRetryOptions } from './interfaces';
 import { HttpClient } from './httpClient';
 import { PlatformNotSupportedError, DistributionNotSupportedError } from './errors';
-
+import * as AsyncRetry from 'async-retry';
```
Comment (anthonydresser) [inline src/serviceDownloadProvider.ts]: use camel case

File: src/serviceDownloadProvider.ts (hunk 11 above)
Comment (ranasaria) [reply src/serviceDownloadProvider.ts]: done. Thanks!

File: src/serviceDownloadProvider.ts [hunk 13]
```diff
@@ -77,35 +78,43 @@ export class ServiceDownloadProvider {
...
+        const urlString = this.getGetDownloadUrl(fileName);
+
+        const pkg: IPackage = {
+            installPath: installDirectory,
+            url: urlString,
+            tmpFile: undefined
+        };
+
+		const existsAsync = promisify(fs.exists);
+		const unlinkAsync = promisify(fs.unlink);
```
Comment (anthonydresser) [inline src/serviceDownloadProvider.ts]: Looks like the formatting is off on this file
